### PR TITLE
improving token config and create bindingJSON utils

### PR DESCRIPTION
**Pull request by huf0813**
Hi, i see a lot of insight from @Endrawan and @AFauzulh for better code. Thx guys, i hope there will be more discussions later

### CODE REVIEW

File: controllers/testing/test_handler.go [hunk 1]
```diff
@@ -31,8 +33,8 @@ type UserTesting struct {
 
 func (cobs *Cobs) Login(c *gin.Context) {
 	var userCobs UserTesting
-	if err := c.BindJSON(&userCobs); err != nil {
-		cobs.Res.CustomResponse(c, "Content-Type", "application/json", "error", "failed when parsing data", http.StatusBadRequest, nil)
+	if err := cobs.Binding.BindingJSON(c, &userCobs); err != nil {
```
Comment (Endrawan) [inline controllers/testing/test_handler.go]: Why not use `ShouldBindJSON()` ?

File: utils/bind/bind.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+func (b *Binding) BindingJSON(c *gin.Context, obj interface{}) error {
+	if err := c.ShouldBindWith(obj, binding.JSON); err != nil {
+		if err := c.AbortWithError(http.StatusBadRequest, err).SetType(gin.ErrorTypeBind); err != nil {
+			return err
+		}
+		return err
+	}
+	return nil
+}
```
Comment (Endrawan) [inline utils/bind/bind.go]: Is this really necessary? I mean gin already provided `ShouldBindJSON()` , which exactly works like this

**Review by Endrawan — CHANGES_REQUESTED** [review]
Thank you for your token improvement @huf0813 . But i still don't understand why you need to create that BindingJSON() method. Maybe you could enlighten me a little?

**Comment by huf0813** [comment]
As you can see on docs. `BindJSON()` by default will implements two methods, they are `ShouldBindWith()` and `AbortWithError()`. If we just implement just `ShouldBindJSON()` (just implementing  `ShouldBindWith()`) then we will missing something important, that is **error handling**. That's why i create utils for those.

docs :
1. [Binding Model](https://github.com/gin-gonic/gin#model-binding-and-validation)
2. [BindJSON()](https://godoc.org/github.com/gin-gonic/gin#Context.BindJSON)
3. [ShouldBindJSON()](https://godoc.org/github.com/gin-gonic/gin#Context.ShouldBindJSON)
4. [ShoulBindWIth()](https://godoc.org/github.com/gin-gonic/gin#Context.ShouldBindWith)

i'll wrap it out, simply like these :
1. `ShouldBindWith() + AbortWithError() = BindJSON()`
2. `ShouldBindWith() + missing error handling = ShouldBindJSON()`

**Comment by Endrawan** [comment]
Okay, but the problem is lies in `c.AbortWithError()` you called in `BindingJSON()`. When you called `c.AbortWithError()`, it will force our server to return a header (in your code its StatusBadRequest). So when we want to change the header to the correct one (for example StatusNotFound), it'll return a warning `Headers were already written. Wanted to override status code 400 with 404`. Not only that, the `c.AbortWithError()` will force the response to text/plain as you can see in this [issue](https://github.com/gin-gonic/gin/issues/633).

So the point of using `ShouldBindJSON()` is to handle the error by ourselves not by the gin itself. Just try to change your code 
```go
if err := cobs.Binding.BindingJSON(c, &userCobs); err != nil {
	cobs.Res.CustomResponse(c, "Content-Type", "application/json", "error", "failed when binding data", http.StatusNotFound, nil)
	return
}
```
to this
```go
if err := c.ShouldBindJSON(&userCobs); err != nil {
	cobs.Res.CustomResponse(c, "Content-Type", "application/json", "error", err.Error(), http.StatusNotFound, nil)
	return
}
```
Then you start your bad request (no body, username with int datatype, etc.) with your postman and see the differences.

**Comment by huf0813** [comment]
Thx, i need to learn more and will delete that later. Don't feel like bothered by me, ok?

**Comment by Endrawan** [comment]
Yeah, chill man. It's just a discussion nothing wrong with it

**Review by Endrawan — APPROVED** [review]
It's all good now. Thank you! @huf0813

**Pull request merged**